python/ray/data/_internal/arrow_serialization.py

```diff
@@ -1,65 +1,118 @@
+# in Arrow < 8.0.0.
+def _register_arrow_json_readoptions_serializer(serialization_context):
     if (
         os.environ.get(
             RAY_DISABLE_CUSTOM_ARROW_JSON_OPTIONS_SERIALIZATION,
             "0",
         )
         == "1"
     ):
-        import logging
-
-        logger = logging.getLogger(__name__)
-        logger.info("Disabling custom Arrow JSON ReadOptions serialization.")
         return
 
-    try:
-        import pyarrow.json as pajson
-    except ModuleNotFoundError:
-        return
+    import pyarrow.json as pajson
 
-    ray.util.register_serializer(
+    serialization_context._register_cloudpickle_serializer(
         pajson.ReadOptions,
-        serializer=lambda opts: (opts.use_threads, opts.block_size),
-        deserializer=lambda args: pajson.ReadOptions(*args),
+        custom_serializer=lambda opts: (opts.use_threads, opts.block_size),
+        custom_deserializer=lambda args: pajson.ReadOptions(*args),
     )
 
 
-def _register_arrow_json_parseoptions_serializer():
-    import ray
-
+def _register_arrow_json_parseoptions_serializer(serialization_context):
     if (
         os.environ.get(
             RAY_DISABLE_CUSTOM_ARROW_JSON_OPTIONS_SERIALIZATION,
             "0",
         )
         == "1"
     ):
-        import logging
-
-        logger = logging.getLogger(__name__)
-        logger.info("Disabling custom Arrow JSON ParseOptions serialization.")
         return
 
-    try:
-        import pyarrow.json as pajson
-    except ModuleNotFoundError:
-        return
+    import pyarrow.json as pajson
 
-    ray.util.register_serializer(
+    serialization_context._register_cloudpickle_serializer(
         pajson.ParseOptions,
-        serializer=lambda opts: (
+        custom_serializer=lambda opts: (
             opts.explicit_schema,
             opts.newlines_in_values,
             opts.unexpected_field_behavior,
         ),
-        deserializer=lambda args: pajson.ParseOptions(*args),
+        custom_deserializer=lambda args: pajson.ParseOptions(*args),
     )
+
+
+# Register custom Arrow data serializer to work around zero-copy slice pickling bug.
+# See https://issues.apache.org/jira/browse/ARROW-10739.
+def _register_arrow_data_serializer(serialization_context):
+    """Custom reducer for Arrow data that works around a zero-copy slicing pickling
+    bug by using the Arrow IPC format for the underlying serialization.
+
+    Background:
+        Arrow has both array-level slicing and buffer-level slicing; both are zero-copy,
+        but the former has a serialization bug where the entire buffer is serialized
+        instead of just the slice, while the latter's serialization works as expected
+        and only serializes the slice of the buffer. I.e., array-level slicing doesn't
+        propagate the slice down to the buffer when serializing the array.
+
+        All that these copy methods do is, at serialization time, take the array-level
```

Comment:
This seems not relevant with the pyarrow ipc approach to handle serialization?

Reply:
Ah yep this is a holdover from the old implementation, let me update this comment!